packages/hatchify-core/src/dataTypes/number.ts

```diff
@@ -0,0 +1,132 @@
+import type {
+  ControlType,
+  HatchifyNumberProps,
+  PartialAttribute,
+} from "../types"
+
+export function number(props?: HatchifyNumberProps): PartialAttribute<number> {
+  const allowNull = props?.required != null ? !props.required : props?.required
+  const control: ControlType = {
+    type: "Number",
+    allowNull,
+    min: props?.min,
+    max: props?.max,
+    primary: props?.primary,
+    step: props?.step,
+  }
+
+  return {
+    name: `number(${props ? JSON.stringify(props) : ""})`,
+    orm: {
+      sequelize: {
+        type: "DOUBLE",
+        typeArgs: [],
+        allowNull,
+        autoIncrement: props?.autoIncrement,
+        primaryKey: props?.primary,
+        ...(props?.min || props?.max
+          ? {
+              validate: {
+                min: props?.min,
+                max: props?.max,
+              },
+            }
+          : {}),
+      },
+    },
+    control,
+    // Passed  - Any crazy value the client might send as a POST or PATCH
+    // Returns - A type the ORM can use
+    // Throws  - If the data is bad ❓
+    // Example : '2023-07-17T01:45:28.778Z' => new Date('2023-07-17T01:45:28.778Z')
+    //         : throw "'4 $core' is not a valid date";
+    setORMPropertyValue: (jsonValue: number | null): number | null => {
+      return clientMutationCoerce(jsonValue, control)
+    },
+
+    // Passed  - Any crazy STRING value the client might send as GET
+    // Returns - A type the ORM can use
+    // Throws  - If the data is bad ❓
+    // Example : ?filter[age]=xyz ... xyz => throw "xyz is not a number";
+    setORMQueryFilterValue: (queryValue: string): number | null => {
+      if (["null", "undefined"].includes(queryValue)) {
+        if (control.allowNull !== false) return null
+        throw new Error("Non-null value is required")
+      }
+
+      if (isNaN(queryValue as unknown as number)) {
+        throw new Error("Provided value is not a number")
+      }
+
+      return clientMutationCoerce(+queryValue, control)
+    },
+
+    // ===== RESPONSE =====
+    // Passed  - A value from the ORM
+    // Returns - A JSON value that can be serialized
+    // Example : new Date() => '2023-07-17T01:45:28.778Z'
+    serializeORMPropertyValue: (ormValue: number | null): number | null => {
+      return clientMutationCoerce(ormValue, control)
+    },
+  }
+}
+
+export function validateStep(value: number, step: number, min = 0): boolean {
+  if (value == null) return false
+
+  let absValue = Math.abs(value)
+  let absStep = Math.abs(step)
+  let absMin = Math.abs(min)
+
+  while (
+    (absValue > 0 && absValue < 1) ||
+    (absStep > 0 && absStep < 1) ||
+    (absMin > 0 && absMin < 1)
+  ) {
+    value *= 10
+    step *= 10
+    min *= 10
+
+    absValue = Math.abs(value)
```

Comment:
might not be worth it but not using Math.abs is 5x slower than  value*value. I did a small benchmark to validate that in my after hours time :) 
[Benchmark link](https://www.measurethat.net/Benchmarks/ShowResult/448487)

Reply:
After second look I think the following is a good solution:
```
function validateStep(value, step, min) {
    if (value < min) {
        return false;
    }
    var diff = (value - min) % step;
    var epsilon = 0.00001; // a small tolerance
    return diff < epsilon || step - diff < epsilon;
}

```